src/darker/__main__.py

```diff
@@ -125,14 +125,42 @@ def format_edited_parts(
                 # 10. A re-formatted Python file which produces an identical AST was
                 #     created successfully - write an updated file or print the diff
                 #     if there were any changes to the original
+                result_str = pyup(result_str, edited_lines, edited_linenums)
+
                 if result_str != worktree_content:
                     # `result_str` is just `chosen_lines` concatenated with newlines.
                     # We need both forms when showing diffs or modifying files.
                     # Pass them both on to avoid back-and-forth conversion.
-                    yield src, worktree_content, result_str, chosen_lines
+                    yield src, worktree_content, result_str, result_str.splitlines()
                 break
 
 
+def pyup(content, edited_lines, edited_linenums):
+
+    from pyupgrade import (
+        _fix_py2_compatible,
+        _fix_tokens,
+        _fix_percent_format,
+        _fix_py3_plus,
+        _fix_py36_plus,
+    )
+
+    min_version = (3, 6)
+    keep_mock = True
+    content = _fix_py2_compatible(content)
+    content = _fix_tokens(content, min_version)
+    content = _fix_percent_format(content)
+    content = _fix_py3_plus(content, min_version, keep_mock)
+    content = _fix_py36_plus(content)
+    content = content.splitlines()
+
+    pyup_opcodes = diff_and_get_opcodes(edited_lines, content)
+    pyupgrade_chunks = list(opcodes_to_chunks(pyup_opcodes, edited_lines, content))
+    pyup_chosen_lines: List[str] = list(choose_lines(pyupgrade_chunks, edited_linenums))
```

Comment:
`edited_linenums` isn't necessarily correct any more at this point. If Black has e.g. divided a long line on multiple lines, every line number after that gets shifted. This would cause wrong pyupgrade changes to be selected and desired ones skipped. This is why changes by all tools (pyupgrade, isort, Black) need to be done first, and only then the diff be calculated between the user's original version and the automatically modified one.

Reply:
Yes, that what I was thinking, or recomputing the diff WRT the revision after each tool.

I think the problem with (all tools) then (compute diff), is you may end up with chunks larger than you wants.

say original

```
a = set((1, 2))
b    =     c
```

user edit
```diff
- b    =     c
+ b    =     d
```

(black + pyupgrade) then (compute diff), will change both lines, which you don't want to apply.
Though, 
(black, diff) should be applied, then (pyupgrade, diff) should not.

Reply:
There's also the additional potential problem of tools which may change the order of lines in code. `isort` is an example of such a tool, and that's why Darker just includes modifications from it as part of the starting point, and compares Black results to that. I wonder if any of the rules in `pyupgrade` could potentially also change the order of lines in the source code?

Here's an illustration of what code reordering would cause if such changes would be attempted to apply to only modified lines:

Let's say I have this module committed to the `HEAD` of my repository:
```python {.lineNo}
1  import  os
2  import black
3  import sys
```

I then manually correct the first line to `import os` without the extra space:
```python
*1  import os
 2  import black
 3  import sys
```

Run isort:
```bash
isort myfile.py
```

Get this as a result:
```python
 1  import os
*2  import sys
*3
*4  import black
```

Git shows the difference:
```
git diff -U0
```
```diff
git diff -U0|cat
diff --git a/imp.py b/imp.py
index 61a152d..1ed0c5e 100644
--- a/imp.py
+++ b/imp.py
@@ -1,2 +1 @@
-import  os
-import black
+import os
@@ -3,0 +3,2 @@ import sys
+
+import black
```

Only the first chunk falls on line 1 (the only edited line) and would be applied by a Darker-like tool. The result would be:
```python
1  import os
2  import sys
```

So I would actually lose my Black import as a result of applying sorted imports partially.

Reply:
from my look at upgrade that does nto seem to be the case, it really appears to do AST node replacements, so I would not be too worried.